distchan.go

```diff
@@ -105,193 +142,196 @@ func (s *Server) Logger() *log.Logger {
 }
 
 func (s *Server) handleOutgoingMessages() {
-	s.WaitUntilReady()
+	var (
+		buf = getBuffer()
+		enc = gob.NewEncoder(&buf)
+	)
 
 	for {
-		x, ok := s.outv.Recv()
+		buf.Reset()
+
+		x, ok := reflect.ValueOf(s.outv).Recv()
 		if !ok {
 			break
 		}
 
-		for {
-			s.mu.RLock()
-			if len(s.connections) == 0 {
-				s.mu.RUnlock()
-				s.WaitUntilReady()
-				s.mu.RLock()
-			}
-
-			if s.currentConn().enc == nil {
-				s.currentConn().enc = gob.NewEncoder(s.currentConn().buf)
-			}
-
-			if err := s.currentConn().enc.EncodeValue(x); err != nil {
-				s.logger.Println(err)
-
-				s.mu.RUnlock()
-				s.mu.Lock()
-				s.connections = append(s.connections[:s.i], s.connections[s.i+1:]...)
-				s.increment()
-				s.mu.Unlock()
-				s.mu.RLock()
-				continue
-			}
-
-			b := s.currentConn().buf.Bytes()
-			s.currentConn().buf.Reset()
+		if err := enc.EncodeValue(x); err != nil {
+			s.logger.Println(err)
+			continue
+		}
 
-			for _, encoder := range s.encoders {
-				b = encoder(b)
+		for _, encoder := range s.encoders {
+			ec := encoder(&buf)
+			buf.Reset()
+			if _, err := io.Copy(&buf, ec); err != nil {
+				s.logger.Panicln(err)
 			}
+		}
 
-			if err := writeChunk(s.currentConn().c, b); err != nil {
-				s.logger.Println(err)
-
-				s.mu.RUnlock()
-				s.mu.Lock()
-				s.connections = append(s.connections[:s.i], s.connections[s.i+1:]...)
-				s.increment()
-				s.mu.Unlock()
-				s.mu.RLock()
-				continue
+		seen := make(map[net.Conn]bool)
+
+		for i := int32(0); i < atomic.LoadInt32(&s.conncnt); {
+			select {
+			// case <-s.closed:
+			// 	break
+			case c, ok := <-s.chbroad:
+				if ok {
+					// we need only different connections
+					if _, ok := seen[c]; !ok {
+						seen[c] = true
+						bts := buf
+						if err := writeChunk(c, &bts, bts.Len()); err != nil {
+							s.logger.Println(err)
+						}
+						i++
+					}
+				}
 			}
-
-			s.mu.RUnlock()
-			s.mu.Lock()
-			s.increment()
-			s.mu.Unlock()
-			break
 		}
 	}
 
 	if err := s.ln.Close(); err != nil {
 		s.logger.Printf("error closing listener: %s\n", err)
 	}
+
+	if buf.Cap() <= 1<<22 {
+		putBuffer(buf)
+	}
 }
 
-func NewClient(conn net.Conn, out, in interface{}) *Client {
+func NewClient(conn net.Conn, out, in interface{}) (*Client, error) {
+	if in != nil && reflect.ValueOf(in).Kind() != reflect.Chan {
+		return nil, ErrorInIsNotChannel
+	}
+	if out != nil && reflect.ValueOf(out).Kind() != reflect.Chan {
+		return nil, ErrorOutIsNotChannel
+	}
 	return &Client{
 		conn:   conn,
-		outv:   reflect.ValueOf(out),
-		inv:    reflect.ValueOf(in),
+		outv:   out,
+		inv:    in,
 		logger: log.New(os.Stdout, "[distchan] ", log.Lshortfile),
 		done:   make(chan struct{}),
-	}
+	}, nil
 }
 
 // Transformer represents a function that does an arbitrary transformation
 // on a piece of data. It's used for defining custom encoders and decoders
 // for modifying how data is sent across the wire.
-type Transformer func([]byte) []byte
+type Transformer func(src io.Reader) io.Reader
```

Comment:
This is an interesting idea.

Reply:
With this you may include any streaming library into chain